File: study/kimi25/polarquant/examples/kv_cache_demo.py
#!/usr/bin/env python3
"""
KV Cache Compression Demo for LLM Inference.

This example demonstrates how PolarQuant can be used to compress
KV (Key-Value) cache entries in transformer-based language models.

In LLM inference, the KV cache stores intermediate attention computations
for previous tokens. For long sequences, this can consume significant memory.
PolarQuant provides near-lossless compression with high compression ratios.
"""

import numpy as np
import time
from polarquant import PolarQuant, PolarQuantConfig


class SimulatedKVCache:
    """Simulates a KV cache for transformer attention."""
    
    def __init__(self, head_dim: int, max_seq_len: int = 1000, use_compression: bool = True):
        """
        Initialize KV cache.
        
        Args:
            head_dim: Dimension of each attention head
            max_seq_len: Maximum sequence length
            use_compression: Whether to use PolarQuant compression
        """
        self.head_dim = head_dim
        self.max_seq_len = max_seq_len
        self.use_compression = use_compression
        
        # Initialize PolarQuant if compression is enabled
        if use_compression:
            config = PolarQuantConfig(
                dimension=head_dim,
                radius_bits=8,
                angle_bits=4,
                seed=42
            )
            self.quantizer = PolarQuant(config)
            self.compression_ratio = self.quantizer.compression_ratio()
        
        # Storage
        self.keys = []
        self.values = []
        self.compressed_keys = []
        self.compressed_values = []
        
        # Statistics
        self.total_original_bytes = 0
        self.total_compressed_bytes = 0
    
    def append(self, key: np.ndarray, value: np.ndarray):
        """Add a new key-value pair to the cache."""
        if len(self.keys) >= self.max_seq_len:
            raise RuntimeError("KV cache is full")
        
        # Normalize (common in attention)
        key = key / (np.linalg.norm(key) + 1e-10)
        value = value / (np.linalg.norm(value) + 1e-10)
        
        # Store original for comparison
        self.keys.append(key.copy())
        self.values.append(value.copy())
        
        # Update original size (float32 = 4 bytes per element)
        self.total_original_bytes += 2 * key.nbytes  # key + value
        
        if self.use_compression:
            # Compress
            start = time.time()
            ck = self.quantizer.compress(key)
            cv = self.quantizer.compress(value)
            compress_time = time.time() - start
            
            self.compressed_keys.append(ck)
            self.compressed_values.append(cv)
            
            # Estimate compressed size
            # radius_idx: 4 bytes (int32)
            # angle_indices: (head_dim - 1) * 4 bytes
            compressed_size = 2 * (4 + (self.head_dim - 1) * 4)
            self.total_compressed_bytes += compressed_size
            
            return compress_time
        
        return 0.0
    
    def get_key(self, idx: int) -> np.ndarray:
        """Get key at index (decompresses if needed)."""
        if self.use_compression:
            return self.quantizer.decompress(self.compressed_keys[idx])
        return self.keys[idx]
    
    def get_value(self, idx: int) -> np.ndarray:
        """Get value at index (decompresses if needed)."""
        if self.use_compression:
            return self.quantizer.decompress(self.compressed_values[idx])
        return self.values[idx]
    
    def compute_attention_scores(self, query: np.ndarray) -> np.ndarray:
        """
        Compute attention scores for a query against all keys.
        
        Args:
            query: Query vector of shape (head_dim,)
            
        Returns:
            Attention scores of shape (seq_len,)
        """
        query = query / (np.linalg.norm(query) + 1e-10)
        
        scores = []
        for i in range(len(self.keys)):
            key = self.get_key(i)
            score = np.dot(query, key)
            scores.append(score)
        
        return np.array(scores)
    
    def get_memory_stats(self) -> dict:
        """Get memory usage statistics."""
        if self.use_compression:
            actual_ratio = self.total_original_bytes / max(self.total_compressed_bytes, 1)
            return {
                'original_bytes': self.total_original_bytes,
                'compressed_bytes': self.total_compressed_bytes,
                'compression_ratio': actual_ratio,
                'space_saved_pct': (1 - 1/actual_ratio) * 100,
            }
        else:
            return {
                'original_bytes': self.total_original_bytes,
                'compressed_bytes': self.total_original_bytes,
                'compression_ratio': 1.0,
                'space_saved_pct': 0.0,
            }


def demo_kv_cache_compression():
    """Demonstrate KV cache compression."""
    print("=" * 70)
    print("KV Cache Compression Demo")
    print("=" * 70)
    
    # Typical LLM parameters
    head_dim = 64  # Common head dimension (e.g., 64, 128)
    seq_len = 100   # Simulate 100 tokens
    
    print(f"\nConfiguration:")
    print(f"  Head dimension: {head_dim}")
    print(f"  Sequence length: {seq_len}")
    print(f"  Quantization: 8-bit radius, 4-bit angles")
    
    # Create caches
    cache_compressed = SimulatedKVCache(head_dim, use_compression=True)
    cache_uncompressed = SimulatedKVCache(head_dim, use_compression=False)
    
    # Simulate generating KV pairs for a sequence
    print(f"\nGenerating {seq_len} KV pairs...")
    
    total_compress_time = 0.0
    
    for i in range(seq_len):
        # Simulate key and value from transformer layer
        key = np.random.randn(head_dim)
        value = np.random.randn(head_dim)
        
        # Add to uncompressed cache
        cache_uncompressed.append(key, value)
        
        # Add to compressed cache
        compress_time = cache_compressed.append(key, value)
        total_compress_time += compress_time
    
    print(f"Compression time: {total_compress_time*1000:.2f} ms")
    print(f"Average per token: {total_compress_time/seq_len*1000:.3f} ms")
    
    # Memory statistics
    print("\nMemory Usage:")
    stats = cache_compressed.get_memory_stats()
    print(f"  Original size: {stats['original_bytes'] / 1024:.2f} KB")
    print(f"  Compressed size: {stats['compressed_bytes'] / 1024:.2f} KB")
    print(f"  Compression ratio: {stats['compression_ratio']:.2f}x")
    print(f"  Space saved: {stats['space_saved_pct']:.1f}%")
    
    # Test attention computation
    print("\nAttention Score Quality:")
    query = np.random.randn(head_dim)
    
    # Compute with uncompressed keys
    scores_uncompressed = cache_uncompressed.compute_attention_scores(query)
    
    # Compute with compressed keys
    start = time.time()
    scores_compressed = cache_compressed.compute_attention_scores(query)
    decompress_time = time.time() - start
    
    print(f"  Decompression time: {decompress_time*1000:.2f} ms")
    
    # Compare scores
    correlation = np.corrcoef(scores_uncompressed, scores_compressed)[0, 1]
    mae = np.mean(np.abs(scores_uncompressed - scores_compressed))
    
    print(f"  Score correlation: {correlation:.4f}")
    print(f"  Mean absolute error: {mae:.6f}")
    
    # Test top-k accuracy
    k = 5
    top_k_uncompressed = set(np.argsort(scores_uncompressed)[-k:])
    top_k_compressed = set(np.argsort(scores_compressed)[-k:])
    top_k_accuracy = len(top_k_uncompressed & top_k_compressed) / k
    
    print(f"  Top-{k} accuracy: {top_k_accuracy*100:.1f}%")


def demo_attention_head_comparison():
    """Compare different attention head dimensions."""
    print("\n" + "=" * 70)
    print("Attention Head Dimension Comparison")
    print("=" * 70)
    
    head_dims = [32, 64, 128, 256]
    seq_len = 50
    
    print(f"\n{'Head Dim':<12} {'Ratio':<10} {'Cosine':<10} {'Memory (KB)':<15}")
    print("-" * 50)
    
    for head_dim in head_dims:
        cache = SimulatedKVCache(head_dim, use_compression=True)
        
        # Populate cache
        for _ in range(seq_len):
            key = np.random.randn(head_dim)
            value = np.random.randn(head_dim)
            cache.append(key, value)
        
        # Test reconstruction quality
        query = np.random.randn(head_dim)
        query = query / np.linalg.norm(query)
        
        # Get original and reconstructed keys
        original_keys = [cache.keys[i] for i in range(seq_len)]
        reconstructed_keys = [cache.get_key(i) for i in range(seq_len)]
        
        # Compute cosine similarities
        cosine_sims = []
        for orig, recon in zip(original_keys, reconstructed_keys):
            sim = np.dot(orig, recon) / (np.linalg.norm(orig) * np.linalg.norm(recon))
            cosine_sims.append(sim)
        
        mean_cosine = np.mean(cosine_sims)
        
        # Get memory stats
        stats = cache.get_memory_stats()
        
        print(f"{head_dim:<12} {stats['compression_ratio']:<10.2f} "
              f"{mean_cosine:<10.4f} {stats['compressed_bytes']/1024:<15.2f}")


def demo_long_sequence():
    """Demonstrate compression on a long sequence."""
    print("\n" + "=" * 70)
    print("Long Sequence Compression (10K tokens)")
    print("=" * 70)
    
    head_dim = 64
    seq_len = 10000
    
    print(f"\nSimulating {seq_len} tokens with head_dim={head_dim}")
    
    cache = SimulatedKVCache(head_dim, max_seq_len=seq_len+100, use_compression=True)
    
    # Generate KV pairs
    for i in range(seq_len):
        key = np.random.randn(head_dim)
        value = np.random.randn(head_dim)
        cache.append(key, value)
        
        if (i + 1) % 2000 == 0:
            print(f"  Processed {i+1} tokens...")
    
    # Memory statistics
    stats = cache.get_memory_stats()
    
    print(f"\nMemory Usage:")
    print(f"  Original: {stats['original_bytes'] / (1024*1024):.2f} MB")
    print(f"  Compressed: {stats['compressed_bytes'] / (1024*1024):.2f} MB")
    print(f"  Compression ratio: {stats['compression_ratio']:.2f}x")
    print(f"  Space saved: {stats['space_saved_pct']:.1f}%")
    
    # Test random access
    print(f"\nRandom Access Test:")
    n_samples = 100
    sample_indices = np.random.choice(seq_len, n_samples, replace=False)
    
    start = time.time()
    for idx in sample_indices:
        _ = cache.get_key(idx)
        _ = cache.get_value(idx)
    elapsed = time.time() - start
    
    print(f"  Accessed {n_samples} random KV pairs in {elapsed*1000:.2f} ms")
    print(f"  Average access time: {elapsed/n_samples*1000:.3f} ms")


def demo_quality_vs_compression():
    """Explore quality vs compression tradeoff."""
    print("\n" + "=" * 70)
    print("Quality vs Compression Tradeoff")
    print("=" * 70)
    
    head_dim = 128
    seq_len = 100
    
    bit_configs = [
        (4, 2, "Low"),
        (6, 3, "Medium-Low"),
        (8, 4, "Medium"),
        (10, 6, "Medium-High"),
        (12, 8, "High"),
    ]
    
    print(f"\n{'Config':<15} {'Radius':<8} {'Angle':<8} {'Ratio':<8} {'Cosine':<10}")
    print("-" * 60)
    
    for r_bits, a_bits, label in bit_configs:
        # Create custom config
        config = PolarQuantConfig(
            dimension=head_dim,
            radius_bits=r_bits,
            angle_bits=a_bits,
            seed=42
        )
        
        cache = SimulatedKVCache(head_dim, use_compression=True)
        cache.quantizer = PolarQuant(config)
        cache.compression_ratio = cache.quantizer.compression_ratio()
        
        # Populate
        for _ in range(seq_len):
            key = np.random.randn(head_dim)
            value = np.random.randn(head_dim)
            cache.append(key, value)
        
        # Test quality
        cosine_sims = []
        for i in range(seq_len):
            orig = cache.keys[i]
            recon = cache.get_key(i)
            sim = np.dot(orig, recon) / (np.linalg.norm(orig) * np.linalg.norm(recon))
            cosine_sims.append(sim)
        
        mean_cosine = np.mean(cosine_sims)
        
        print(f"{label:<15} {r_bits:<8} {a_bits:<8} "
              f"{cache.compression_ratio:<8.2f} {mean_cosine:<10.4f}")


def main():
    """Run all demos."""
    print("\n")
    print("*" * 70)
    print("*" + " " * 68 + "*")
    print("*" + "  PolarQuant: KV Cache Compression Demo".center(68) + "*")
    print("*" + " " * 68 + "*")
    print("*" * 70)
    
    np.random.seed(42)
    
    demo_kv_cache_compression()
    demo_attention_head_comparison()
    demo_long_sequence()
    demo_quality_vs_compression()
    
    print("\n" + "=" * 70)
    print("All demos completed!")
    print("=" * 70)


if __name__ == "__main__":
    main()

File: study/glm51/examples/kv_cache_demo.rs
// PolarQuant KV Cache 压缩演示
//
// 模拟 LLM 推理中的 KV Cache 压缩场景，展示 PolarQuant 的：
// 1. 压缩效率（压缩比、内存节省）
// 2. 重建质量（余弦相似度、注意力分数保持）
// 3. 不同维度/位宽下的性能对比
// 4. 长序列（10K tokens）的可扩展性

use polarquant::{CompressedVector, PolarQuant, PolarQuantConfig};
use rand::Rng;
use std::time::Instant;

/// 向量归一化到单位球面
/// 
/// # 参数
/// - `x`: 输入向量
/// 
/// # 返回值
/// 归一化后的向量（L2范数为1）
fn normalize_vector(x: &[f64]) -> Vec<f64> {
    let norm: f64 = x.iter().map(|v| v * v).sum::<f64>().sqrt();
    if norm < 1e-10 {
        return x.to_vec();
    }
    x.iter().map(|&v| v / norm).collect()
}

/// 向量点积
/// 
/// # 参数
/// - `a`: 第一个向量
/// - `b`: 第二个向量
/// 
/// # 返回值
/// 两个向量的点积值
fn dot(a: &[f64], b: &[f64]) -> f64 {
    a.iter().zip(b.iter()).map(|(&x, &y)| x * y).sum()
}

/// 余弦相似度
/// 
/// # 参数
/// - `a`: 第一个向量
/// - `b`: 第二个向量
/// 
/// # 返回值
/// 余弦相似度值，范围[-1, 1]
fn cosine_similarity(a: &[f64], b: &[f64]) -> f64 {
    let na: f64 = a.iter().map(|v| v * v).sum::<f64>().sqrt();
    let nb: f64 = b.iter().map(|v| v * v).sum::<f64>().sqrt();
    if na < 1e-10 || nb < 1e-10 {
        return 0.0;
    }
    dot(a, b) / (na * nb)
}

/// 模拟 KV Cache 结构
///
/// 支持：
/// - 原始模式：存储完整 f64 向量
/// - 压缩模式：使用 PolarQuant 压缩后存储
struct SimulatedKVCache {
    head_dim: usize,               // 注意力头维度
    use_compression: bool,         // 是否启用压缩
    quantizer: Option<PolarQuant>, // 量化器实例
    keys: Vec<Vec<f64>>,           // 原始 Key 向量（用于质量评估）
    values: Vec<Vec<f64>>,         // 原始 Value 向量
    compressed_keys: Vec<CompressedVector>,   // 压缩后的 Key
    compressed_values: Vec<CompressedVector>, // 压缩后的 Value
    total_original_bytes: usize,   // 原始数据总字节数
    total_compressed_bytes: usize, // 压缩后数据总字节数
}

impl SimulatedKVCache {
    /// 创建 KV Cache 实例
    /// 
    /// # 参数
    /// - `head_dim`: 注意力头维度
    /// - `use_compression`: 是否启用压缩
    fn new(head_dim: usize, use_compression: bool) -> Self {
        let quantizer = if use_compression {
            let config = PolarQuantConfig::builder(head_dim)
                .radius_bits(8)
                .angle_bits(4)
                .seed(42)
                .build()
                .unwrap();
            Some(PolarQuant::new(config).unwrap())
        } else {
            None
        };

        Self {
            head_dim,
            use_compression,
            quantizer,
            keys: Vec::new(),
            values: Vec::new(),
            compressed_keys: Vec::new(),
            compressed_values: Vec::new(),
            total_original_bytes: 0,
            total_compressed_bytes: 0,
        }
    }

    /// 追加一个 token 的 KV 对
    ///
    /// # 参数
    /// - `key`: Key 向量
    /// - `value`: Value 向量
    /// 
    /// # 返回值
    /// 压缩耗时（秒）
    fn append(&mut self, key: &[f64], value: &[f64]) -> f64 {
        let key_norm = normalize_vector(key);
        let value_norm = normalize_vector(value);

        // 保存原始向量（用于质量评估）
        self.keys.push(key_norm.clone());
        self.values.push(value_norm.clone());

        // 原始大小：2 个 head_dim 维 f64 向量
        self.total_original_bytes += 2 * self.head_dim * 8;

        if self.use_compression {
            let start = Instant::now();

            // 压缩 Key 和 Value
            let ck = self.quantizer.as_ref().unwrap().compress(&key_norm).unwrap();
            let cv = self.quantizer.as_ref().unwrap().compress(&value_norm).unwrap();

            let elapsed = start.elapsed().as_secs_f64();

            // 压缩后大小估算：radius(4B) + (d-1) * angle(4B)，每个 KV 对
            let compressed_size = 2 * (4 + (self.head_dim - 1) * 4);
            self.total_compressed_bytes += compressed_size;

            self.compressed_keys.push(ck);
            self.compressed_values.push(cv);

            elapsed
        } else {
            // 未压缩模式：与原始大小相同
            self.total_compressed_bytes += 2 * self.head_dim * 8;
            0.0
        }
    }

    /// 获取第 idx 个 Key 向量（压缩模式自动解压）
    /// 
    /// # 参数
    /// - `idx`: 索引
    fn get_key(&self, idx: usize) -> Vec<f64> {
        if self.use_compression {
            self.quantizer
                .as_ref()
                .unwrap()
                .decompress(&self.compressed_keys[idx])
        } else {
            self.keys[idx].clone()
        }
    }

    /// 获取第 idx 个 Value 向量（压缩模式自动解压）
    /// 
    /// # 参数
    /// - `idx`: 索引
    fn get_value(&self, idx: usize) -> Vec<f64> {
        if self.use_compression {
            self.quantizer
                .as_ref()
                .unwrap()
                .decompress(&self.compressed_values[idx])
        } else {
            self.values[idx].clone()
        }
    }

    /// 计算注意力分数（query 与所有 keys 的点积）
    /// 
    /// # 参数
    /// - `query`: 查询向量
    /// 
    /// # 返回值
    /// 注意力分数列表
    fn compute_attention_scores(&self, query: &[f64]) -> Vec<f64> {
        let query_norm = normalize_vector(query);
        (0..self.keys.len())
            .map(|i| {
                let key = self.get_key(i);
                dot(&query_norm, &key)
            })
            .collect()
    }

    /// 计算实际压缩比
    /// 
    /// # 返回值
    /// 压缩比 = 原始大小 / 压缩后大小
    fn compression_ratio(&self) -> f64 {
        if self.total_compressed_bytes == 0 {
            return 1.0;
        }
        self.total_original_bytes as f64 / self.total_compressed_bytes as f64
    }

    /// 计算节省空间百分比
    /// 
    /// # 返回值
    /// 节省空间百分比
    fn space_saved_pct(&self) -> f64 {
        let ratio = self.compression_ratio();
        if ratio <= 1.0 {
            return 0.0;
        }
        (1.0 - 1.0 / ratio) * 100.0
    }
}

/// 演示 1: KV Cache 基础压缩
///
/// 模拟 100 个 token 的 KV Cache，评估：
/// - 压缩耗时
/// - 内存节省
/// - 注意力分数保持度
/// - Top-K 检索准确率
fn demo_kv_cache_compression() {
    println!("{}", "=".repeat(70));
    println!("KV Cache 压缩基础演示");
    println!("{}", "=".repeat(70));

    let head_dim = 64;
    let seq_len = 100;

    println!("\n配置:");
    println!("  注意力头维度: {}", head_dim);
    println!("  序列长度: {}", seq_len);
    println!("  量化配置: 8-bit 半径, 4-bit 角度");

    let mut cache_compressed = SimulatedKVCache::new(head_dim, true);
    let mut cache_uncompressed = SimulatedKVCache::new(head_dim, false);

    let mut rng = rand::thread_rng();
    let mut total_compress_time = 0.0f64;

    println!("\n生成 {} 个 KV 对...", seq_len);

    for _ in 0..seq_len {
        let key: Vec<f64> = (0..head_dim).map(|_| rng.gen::<f64>() * 2.0 - 1.0).collect();
        let value: Vec<f64> = (0..head_dim).map(|_| rng.gen::<f64>() * 2.0 - 1.0).collect();

        cache_uncompressed.append(&key, &value);
        total_compress_time += cache_compressed.append(&key, &value);
    }

    println!("压缩耗时: {:.2} ms", total_compress_time * 1000.0);
    println!(
        "平均每 token: {:.3} ms",
        total_compress_time / seq_len as f64 * 1000.0
    );

    println!("\n内存使用:");
    println!(
        "  原始大小: {:.2} KB",
        cache_compressed.total_original_bytes as f64 / 1024.0
    );
    println!(
        "  压缩后大小: {:.2} KB",
        cache_compressed.total_compressed_bytes as f64 / 1024.0
    );
    println!(
        "  实际压缩比: {:.2}x (u32 索引存储)",
        cache_compressed.compression_ratio()
    );
    let theoretical_ratio = cache_compressed.quantizer.as_ref().unwrap().compression_ratio();
    println!(
        "  理论压缩比 (bit-packed): {:.2}x",
        theoretical_ratio
    );
    println!("  节省空间: {:.1}%", cache_compressed.space_saved_pct());

    // 注意力分数质量评估
    println!("\n注意力分数质量:");
    let query: Vec<f64> = (0..head_dim).map(|_| rng.gen::<f64>() * 2.0 - 1.0).collect();

    let scores_uncompressed = cache_uncompressed.compute_attention_scores(&query);

    let start = Instant::now();
    let scores_compressed = cache_compressed.compute_attention_scores(&query);
    let decompress_time = start.elapsed().as_secs_f64();

    println!(
        "  解压耗时: {:.2} ms",
        decompress_time * 1000.0
    );

    // 计算 Pearson 相关系数
    let n = scores_uncompressed.len().min(scores_compressed.len());
    let mean_uncompressed: f64 = scores_uncompressed.iter().sum::<f64>() / n as f64;
    let mean_compressed: f64 = scores_compressed.iter().sum::<f64>() / n as f64;

    let cov: f64 = scores_uncompressed
        .iter()
        .zip(scores_compressed.iter())
        .map(|(&a, &b)| (a - mean_uncompressed) * (b - mean_compressed))
        .sum::<f64>()
        / n as f64;
    let var_u: f64 = scores_uncompressed
        .iter()
        .map(|a| (a - mean_uncompressed).powi(2))
        .sum::<f64>()
        / n as f64;
    let var_c: f64 = scores_compressed
        .iter()
        .map(|a| (a - mean_compressed).powi(2))
        .sum::<f64>()
        / n as f64;
    let correlation = cov / (var_u.sqrt() * var_c.sqrt());

    // 平均绝对误差
    let mae: f64 = scores_uncompressed
        .iter()
        .zip(scores_compressed.iter())
        .map(|(&a, &b)| (a - b).abs())
        .sum::<f64>()
        / n as f64;

    println!("  分数相关性 (Pearson): {:.4}", correlation);
    println!("  平均绝对误差: {:.6}", mae);

    // Top-5 检索准确率
    let mut indexed_scores: Vec<(usize, f64)> = scores_uncompressed
        .iter()
        .enumerate()
        .map(|(i, &s)| (i, s))
        .collect();
    indexed_scores.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
    let top_k_uncompressed: Vec<usize> = indexed_scores.iter().rev().take(5).map(|(i, _)| *i).collect();

    let mut indexed_scores_c: Vec<(usize, f64)> = scores_compressed
        .iter()
        .enumerate()
        .map(|(i, &s)| (i, s))
        .collect();
    indexed_scores_c.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
    let top_k_compressed: Vec<usize> = indexed_scores_c.iter().rev().take(5).map(|(i, _)| *i).collect();

    let top_k_overlap = top_k_uncompressed
        .iter()
        .filter(|i| top_k_compressed.contains(i))
        .count();
    let top_k_accuracy = top_k_overlap as f64 / 5.0;

    println!("  Top-5 检索准确率: {:.1}%", top_k_accuracy * 100.0);
}

/// 演示 2: 不同注意力头维度的压缩效果对比
///
/// 维度越高，Beta 分布越集中，量化精度越高
fn demo_attention_head_comparison() {
    println!("\n{}", "=".repeat(70));
    println!("注意力头维度对比");
    println!("{}", "=".repeat(70));

    let head_dims = [32, 64, 128, 256];
    let seq_len = 50;

    println!(
        "\n{:<12} {:<10} {:<10} {:<15}",
        "头维度", "压缩比", "余弦相似度", "内存 (KB)"
    );
    println!("{}", "-".repeat(50));

    let mut rng = rand::thread_rng();

    for &head_dim in &head_dims {
        let mut cache = SimulatedKVCache::new(head_dim, true);

        for _ in 0..seq_len {
            let key: Vec<f64> = (0..head_dim).map(|_| rng.gen::<f64>() * 2.0 - 1.0).collect();
            let value: Vec<f64> = (0..head_dim).map(|_| rng.gen::<f64>() * 2.0 - 1.0).collect();
            cache.append(&key, &value);
        }

        // 计算重建质量
        let mut cosine_sims = Vec::new();
        for i in 0..seq_len {
            let orig = &cache.keys[i];
            let recon = cache.get_key(i);
            cosine_sims.push(cosine_similarity(orig, &recon));
        }
        let mean_cosine = cosine_sims.iter().sum::<f64>() / cosine_sims.len() as f64;

        println!(
            "{:<12} {:<10.2} {:<10.4} {:<15.2}",
            head_dim,
            cache.compression_ratio(),
            mean_cosine,
            cache.total_compressed_bytes as f64 / 1024.0
        );
    }
}

/// 演示 3: 长序列压缩（10K tokens）
///
/// 测试大规模 KV Cache 的压缩性能和随机访问速度
fn demo_long_sequence() {
    println!("\n{}", "=".repeat(70));
    println!("长序列压缩 (10K tokens)");
    println!("{}", "=".repeat(70));

    let head_dim = 64;
    let seq_len = 10000;

    println!("\n模拟 {} 个 tokens，head_dim={}", seq_len, head_dim);

    let mut cache = SimulatedKVCache::new(head_dim, true);
    let mut rng = rand::thread_rng();

    for i in 0..seq_len {
        let key: Vec<f64> = (0..head_dim).map(|_| rng.gen::<f64>() * 2.0 - 1.0).collect();
        let value: Vec<f64> = (0..head_dim).map(|_| rng.gen::<f64>() * 2.0 - 1.0).collect();
        cache.append(&key, &value);

        if (i + 1) % 2000 == 0 {
            println!("  已处理 {} tokens...", i + 1);
        }
    }

    println!("\n内存使用:");
    println!(
        "  原始: {:.2} MB",
        cache.total_original_bytes as f64 / (1024.0 * 1024.0)
    );
    println!(
        "  压缩后: {:.2} MB",
        cache.total_compressed_bytes as f64 / (1024.0 * 1024.0)
    );
    println!(
        "  实际压缩比: {:.2}x (u32 索引存储)",
        cache.compression_ratio()
    );
    let theoretical_ratio = cache.quantizer.as_ref().unwrap().compression_ratio();
    println!(
        "  理论压缩比 (bit-packed): {:.2}x",
        theoretical_ratio
    );
    println!("  节省空间: {:.1}%", cache.space_saved_pct());

    // 随机访问性能测试
    println!("\n随机访问测试:");
    let n_samples = 100;
    let mut sample_indices: Vec<usize> = Vec::new();
    for _ in 0..n_samples {
        sample_indices.push(rng.gen_range(0..seq_len));
    }

    let start = Instant::now();
    for &idx in &sample_indices {
        let _ = cache.get_key(idx);
        let _ = cache.get_value(idx);
    }
    let elapsed = start.elapsed().as_secs_f64();

    println!(
        "  访问 {} 个随机 KV 对耗时 {:.2} ms",
        n_samples,
        elapsed * 1000.0
    );
    println!(
        "  平均访问时间: {:.3} ms",
        elapsed / n_samples as f64 * 1000.0
    );
}

/// 演示 4: 质量 vs 压缩率权衡
///
/// 不同量化位宽下的压缩率和重建质量对比
fn demo_quality_vs_compression() {
    println!("\n{}", "=".repeat(70));
    println!("质量 vs 压缩率权衡");
    println!("{}", "=".repeat(70));

    let head_dim = 128;
    let seq_len = 100;

    let bit_configs = [
        (4, 2, "低质量"),
        (6, 3, "中低质量"),
        (8, 4, "中等质量"),
        (10, 6, "中高质量"),
        (12, 8, "高质量"),
    ];

    println!(
        "\n{:<15} {:<8} {:<8} {:<8} {:<10}",
        "配置", "半径位", "角度位", "压缩比", "余弦"
    );
    println!("{}", "-".repeat(60));

    let mut rng = rand::thread_rng();

    for (r_bits, a_bits, label) in bit_configs {
        let config = PolarQuantConfig::builder(head_dim)
            .radius_bits(r_bits)
            .angle_bits(a_bits)
            .seed(42)
            .build()
            .unwrap();
        let pq = PolarQuant::new(config).unwrap();

        let mut keys = Vec::new();
        let mut compressed_keys = Vec::new();

        for _ in 0..seq_len {
            let key: Vec<f64> = (0..head_dim).map(|_| rng.gen::<f64>() * 2.0 - 1.0).collect();
            let key_norm = normalize_vector(&key);
            keys.push(key_norm.clone());
            compressed_keys.push(pq.compress(&key_norm).unwrap());
        }

        // 计算平均余弦相似度
        let mut cosine_sims = Vec::new();
        for i in 0..seq_len {
            let recon = pq.decompress(&compressed_keys[i]);
            cosine_sims.push(cosine_similarity(&keys[i], &recon));
        }
        let mean_cosine = cosine_sims.iter().sum::<f64>() / cosine_sims.len() as f64;

        println!(
            "{:<15} {:<8} {:<8} {:<8.2} {:<10.4}",
            label,
            r_bits,
            a_bits,
            pq.compression_ratio(),
            mean_cosine
        );
    }
}

/// 主函数：运行所有演示
fn main() {
    println!();
    println!("{}", "*".repeat(70));
    println!("*{}*", " ".repeat(68));
    let title = "  PolarQuant: KV Cache 压缩演示 (Rust)";
    let padded: String = title
        .chars()
        .chain(std::iter::repeat(' '))
        .take(68)
        .collect();
    println!("*{}*", padded);
    println!("*{}*", " ".repeat(68));
    println!("{}", "*".repeat(70));

    demo_kv_cache_compression();
    demo_attention_head_comparison();
    demo_long_sequence();
    demo_quality_vs_compression();

    println!("\n{}", "=".repeat(70));
    println!("所有演示完成！");
    println!("{}", "=".repeat(70));
}
